plugins/search-backend-node/src/collators/NewlineDelimitedJsonCollatorFactory.ts

```diff
@@ -0,0 +1,147 @@
+import { Logger } from 'winston';
+
+/**
+ * @beta
+ */
+export type NewlineDelimitedJsonCollatorFactoryOptions = {
+  type: string;
+  searchPattern: string;
+  reader: UrlReader;
+  logger: Logger;
+  visibilityPermission?: Permission;
+};
+
+/**
+ * Factory class producing a collator that can be used to index documents
+ * sourced from the latest newline delimited JSON file matching a given search
+ * pattern. "Latest" is determined by the name of the file (last alphabetically
+ * is considered latest).
+ *
+ * @remarks
+ * The reader provided must implement the `search()` method as well as the
+ * `readUrl` method whose response includes the `stream()` method. Naturally,
+ * the reader must also be configured to understand the given search pattern.
+ *
+ * @example
+ * Here's an example configuration using Google Cloud Storage, which would
+ * return the latest file under the `bucket` GCS bucket with files like
+ * `xyz-2021.ndjson` or `xyz-2022.ndjson`.
+ * ```ts
+ * indexBuilder.addCollator({
+ *   schedule,
+ *   factory: NewlineDelimitedJsonCollatorFactory.fromConfig(env.config, {
+ *     type: 'techdocs',
+ *     searchPattern: 'https://storage.cloud.google.com/bucket/xyz-*',
+ *     reader: env.reader,
+ *     logger: env.logger,
+ *   })
+ * });
+ * ```
+ *
+ * @beta
+ */
+export class NewlineDelimitedJsonCollatorFactory
+  implements DocumentCollatorFactory
+{
+  /** {@inheritDoc @backstage/plugin-search-common#DocumentCollatorFactory."type"} */
+  readonly type: string;
+
+  /** {@inheritDoc @backstage/plugin-search-common#DocumentCollatorFactory.visibilityPermission} */
+  public readonly visibilityPermission: Permission | undefined;
+
+  private constructor(
+    type: string,
+    private readonly searchPattern: string,
+    private readonly reader: UrlReader,
+    private readonly logger: Logger,
+    visibilityPermission: Permission | undefined,
+  ) {
+    this.type = type;
+    this.visibilityPermission = visibilityPermission;
+  }
+
+  /**
+   * Returns a NewlineDelimitedJsonCollatorFactory instance from configuration
+   * and a set of options.
+   */
+  static fromConfig(
+    _config: Config,
+    options: NewlineDelimitedJsonCollatorFactoryOptions,
+  ): NewlineDelimitedJsonCollatorFactory {
+    return new NewlineDelimitedJsonCollatorFactory(
+      options.type,
+      options.searchPattern,
+      options.reader,
+      options.logger,
+      options.visibilityPermission,
+    );
+  }
+
+  /**
+   * Returns the "latest" URL for the given search pattern (e.g. the one at the
+   * end of the list, sorted alphabetically).
+   */
+  private async lastUrl(): Promise<string | undefined> {
+    try {
+      // Search for files matching the given pattern, then sort/reverse. The
+      // first item in the list will be the "latest" file.
+      this.logger.info(
+        `Attempting to find latest .ndjson matching ${this.searchPattern}`,
+      );
+      const { files } = await this.reader.search(this.searchPattern);
```

Comment:
It would be nice!  But unfortunately not.  I think the `UrlReader` ends up being used for quite a few different purposes, so the API surface area is very limited (basically just a `glob`).

Having many URLs/files returned could definitely be a problem, but there might be ways to get around it by passing just the right search pattern.